### ISSUE TITLE
Non-homogeneous Dirichlet boundary condition on only one part of the complex function  

### ISSUE THREAD

**Issue by MarijaI** [mfem/mfem#3069]
Hi everyone, 

I am new to MFEM and I am trying to add three different boundary conditions to the faces of the waveguide:
- the entrance of the waveguide:  
      Re(a) = n x (Curl u); 
      Im(a) = beta * n x (n x u);
      Re(b) = 0;
      Im(b) = f; 
- the exit of the waveguide: 
      Re(a)  = n x (Curl u);
      Im(a)  =  beta * n x (n x u);
      Re(b) = 0;
      Im(b) = 0; 
- the rest of the waveguide should be the perfect electrical conductor: n x u = 0.
beta - constant,
f - vector function, 
i - imaginary unit,
n - normal vector
u - vector function
I don't know how to add the condition n x (n x u) = f to the imaginary part of the complex function.
This is my code. 
....
 FiniteElementCollection* fec = new ND_FECollection(order, dim);   // Nedelec Finite Elements
 FiniteElementSpace* fespace = new FiniteElementSpace(mesh, fec);
 ComplexGridFunction u(fespace);    // solution
 u = 0.0;
 
ComplexLinearForm b(fespace, conv);
 b.Vector::operator = (0.0);
 b.AddBdrFaceIntegrator(NULL, new VectorFEDomainLFIntegrator(f), WaveguideEntrance); 
 b.Assemble();

SesquilinearForm* a = new SesquilinearForm(fespace, conv);
a->AddDomainIntegrator(new CurlCurlIntegrator(one), NULL);  
a->AddBdrFaceIntegrator(new CurlCurlIntegrator(one), ____________ , WaveguideEntrance);
a->AddBdrFaceIntegrator(new CurlCurlIntegrator(one), ___________, WaveguideExit);   
a->AddBdrFaceIntegrator(new MixedVectorWeakCurlIntegrator(one), _________, WaveguidePEC);                             

Is it possible to add these conditions in the weak form?

Thank you in advance,
Marija

**Comment by MarijaI** [comment]
Hi @mlstowell,
Thank you very much for the comments and help.

> Firstly, the complex linear system support is still quite limited. There are a few convenience methods for the most basic problem configurations but more elaborate scenarios may require manipulating the `LinearForm` and `BilinearForm` objects stored inside the `ComplexLinearForm` and `SesquilinearForm` objects. We can try to offer advice on how to accomplish this if needed.

I would love to try to solve the problem by manipulating with the Linear and Bilinear objects stored inside the ComplexLinearForm and SesquilinearForm objects. If you can point me to some literature (example), that would be very helpful. 

>Next, the AddBdrFaceIntegrator methods are used with Discontinuous Galerkin basis functions rather than Nedelec so I don't believe these will help you. Unfortunately, it doesn't look like our documentation makes this clear. I apologize for that. The corresponding methods for Nedelec, and most other basis function types, would be AddBoundaryIntegrator. These are useful for applying non-homogeneous Neumann boundary to LinearForm objects or impedance boundary conditions to BilinearForm objects.

Ok, that's my bad. Thank you for pointed that out. 

>Finally, from your initial question I'm still a little confused about the boundary conditions that you need to apply at the ends of the waveguide. Normally I would not constrain the real and imaginary parts separately but this could be possible. Can you explain these boundary conditions in more detail? As the web documentation mentioned above describes we normally consider three types of boundary conditions; n x (n x u) = n x f, n x Curl u = n x g, or n x Curl u + gamma n x u = n x f. Do your boundary conditions fit one of these patterns or are they more complicated?

Actually, boundary do fits to the Real and Imaginary part of the equations. 
I have similar boundary conditions:
1. n x Curl u + i * beta * n x (n x u) = i*f 
2. n x Curl u + i * beta * n x (n x u) = 0
3. n x (n x u) = 0
i-imaginary unit, beta = const, f - function, n-unit vector normal to boundary

The third condition is Dirichlet  (Essential) boundary condition for the ND. Is it possible to assign that condition to the specific boundary? If I understood correctly, we don't assigned it explicitly.

I am attaching the original problem
[model.pdf](https://github.com/mfem/mfem/files/8919287/model.pdf)
Maybe it will be more clear from the model. Any help or guidance will be mode than welcome.  

Best regards, Marija

**Comment by MarijaI** [comment]
Hi @psocratis 
Thanx for your help. This helps a lot. 

Still, I have some questions. 
Can you help me with the coefficients because I don't have the exact solution?
I am not sure if I understand it well.
In my problem, the WaveguideEntrance looks like this: 
n x curl E + i beta n x (n x E) = -2i beta E0, where E0 is a function E0=E0(x).
So, for the WaveguideEntrance I should use:
-n x curl E:   
b.AddBoundaryIntegrator(new VectorFEBoundaryTangentLFIntegrator(c1_Re), NULL, WaveguideEntrance);
-i beta n x (n x E):
b.AddBoundaryIntegrator(NULL, new VectorFEDomainLFIntegrator(c2_Im), WaveguideEntrance);
-2i beta E0
how can I add this part to the linear form b?

I have noticed that you calculate coefficients c1_Re and c2_Im using the exact value of the E, but I don't have the exact value of the E. 
Also, you have this conditions both for the Real and Imaginary part. Should I also have them both in the real and imaginary part of the linear form?

Thank you in advance for all the answers and please sorry for bothering. I really want to learn and to understand it well.
Best, Marija

**Comment by Kuloss** [comment]
Hey, @MarijaI, I find we may have the same problem about the boundary condition. are you solved? Do you meet the problem when using the `a.AddBoundaryIntegrator(NULL, new VectorFEMassIntegrator(om),imp_bdr)` to express $\hat{n} \times\left ( jk_{0} \hat{n}\times\overrightarrow{u}  \right )$ , because I found if  Sommerfeld was roughly to setting as that, RCS finale calculated is not going right.

**Comment by MarijaI** [comment]
Hi @Kullos, we are solving the similar problem. From the same literature I am solving problem described with 9.3.37, 9.3.38 and 9.3.39. 
![image](https://user-images.githubusercontent.com/46717785/183689085-e10a2e42-3182-4150-bda2-3060e1f65195.png)
@psocratis helped a lot with his code (with maxwell_impedanceBC.cpp). 
I will explain you how I coded my problem (still I am not sure if it is the correct way). 
I wrote weak formulations for all three equations and got this:
![image](https://user-images.githubusercontent.com/46717785/183696963-716e92e7-4e4a-46ed-ad85-c03240289632.png)
Therefore, my problem looks like this:
![image](https://user-images.githubusercontent.com/46717785/183697417-9f3cb80a-8ad8-4622-8c65-be74b8048d89.png)


```
ComplexGridFunction u(fespace);
u.real() = 0.0;
u.imag() = 0.0;
u.ProjectBdrCoefficientTangent(oneVecCoef, oneVecCoef, ess_bdr);  // oneVecCoef = [0,0,1]

 ComplexLinearForm* b = new ComplexLinearForm(fespace, conv);
 b->Vector::operator = (0.0);

VectorFunctionCoefficient E_Re(dim, E_inc_Re);
VectorFunctionCoefficient E_Im(dim, E_inc_Im);

ScalarVectorProductCoefficient c0_Re(-2. * beta, E_Im);
ScalarVectorProductCoefficient c0_Im(2. * beta, E_Re);
    
b->AddBoundaryIntegrator(new VectorFEDomainLFIntegrator(c0_Re), new VectorFEDomainLFIntegrator(c0_Im), Gamma1);  
b->Assemble();

ConstantCoefficient muinv(1 / mur);
ConstantCoefficient minus_k0_k0_epsilonr(-k0 * k0 * epsilonr);

SesquilinearForm* a = new SesquilinearForm(fespace, conv);                       
a->AddDomainIntegrator(new CurlCurlIntegrator(muinv), NULL);                
a->AddDomainIntegrator(new VectorFEMassIntegrator(minus_k0_k0_epsilonr), NULL);  
ConstantCoefficient om(beta / mur);
a->AddBoundaryIntegrator(NULL, new VectorFEMassIntegrator(om), Gamma1);
a->AddBoundaryIntegrator(NULL, new VectorFEMassIntegrator(om), Gamma2);
a->Assemble();
```
I am not sure if I choose the following integrators correct since I have to calculate integrals over the (n \times v)(n times E). For Gamma3 I don't need integral since that the conditions on the Gamma3 are n \times E = 0.
```
a->AddBoundaryIntegrator(NULL, new VectorFEMassIntegrator(om), Gamma1);
a->AddBoundaryIntegrator(NULL, new VectorFEMassIntegrator(om), Gamma2); 
```

**Comment by MarijaI** [comment]
@psocratis sorry for not answering earlier. I was hopping that I will solve problem by myself. 
Actually, I thought that I was. But, since @Kuloss asked the similar question I was wondering if you can just take a look on the previous comment and on the code I attached. 

I do have analytic expression for the `E_inc` (that is the way how I calculate `E_inc_Re` and `E_inc_Im`). I thought that in the `SesquilinearForm` form we have to separate real and imaginary parts of the equation and since `E_inc` appears only in the imag part of the equation, I am confused why we put it in both. 
Let me rephrase the question. Since E is a complex function, that would mean that I should have both a real and an imaginary part everywhere or not?

Thank you for everything. Best, Marija

**Comment by psocratis** [comment]
Hi @MarijaI, 

What do you mean `E_inc` appears only in the imaginary part? Isn't `E_inc` complex itself? If it is, then you have to separate its real and imaginary part, do the appropriate algebra (multiply by $-2j\beta$ for example) and at the end, you would get a complex expression with both real and imaginary parts. I am not sure if I am missing something, but I do see both real and imaginary parts being non-zero in general. 

Thanks, 

Socratis
